starboard/shared/modular/starboard_layer_posix_socket_abi_wrappers.cc

```diff
@@ -77,27 +79,110 @@ SB_EXPORT int __abi_wrap_connect(int sockfd,
+                                     struct musl_addrinfo** res) {
+  if (res == nullptr)
+    return;
+  // musl addrinfo definition might differ from platform definition.
+  // So we need to do a manual conversion to avoid header mismatches.
+  struct addrinfo new_hints = {0};
 
-#if SB_HAS_QUIRK(SOCKADDR_WITH_LENGTH)
-  struct addrinfo* ai = *res;
+  if (hints != nullptr) {
+    if (hints->ai_flags != 0) {
+      if (hints->ai_flags & MUSL_AI_PASSIVE)
+        new_hints.ai_flags |= AI_PASSIVE;
+      if (hints->ai_flags & MUSL_AI_CANONNAME)
+        new_hints.ai_flags |= AI_CANONNAME;
+      if (hints->ai_flags & MUSL_AI_NUMERICHOST)
+        new_hints.ai_flags |= AI_NUMERICHOST;
+      if (hints->ai_flags & MUSL_AI_V4MAPPED)
+        new_hints.ai_flags |= AI_V4MAPPED;
+      if (hints->ai_flags & MUSL_AI_ALL)
+        new_hints.ai_flags |= AI_ALL;
+      if (hints->ai_flags & MUSL_AI_ADDRCONFIG)
+        new_hints.ai_flags |= AI_ADDRCONFIG;
+      if (hints->ai_flags & MUSL_AI_NUMERICSERV)
+        new_hints.ai_flags |= AI_NUMERICSERV;
+    }
+
+    if (hints->ai_socktype == MUSL_SOCK_STREAM)
+      new_hints.ai_socktype = SOCK_STREAM;
+    else if (hints->ai_socktype == MUSL_SOCK_DGRAM)
+      new_hints.ai_socktype = SOCK_DGRAM;
+    else if (hints->ai_socktype == MUSL_SOCK_RAW)
+      new_hints.ai_socktype = SOCK_RAW;
+
+    if (hints->ai_protocol == MUSL_IPPROTO_TCP)
+      new_hints.ai_protocol = IPPROTO_TCP;
+    else if (hints->ai_protocol == MUSL_IPPROTO_UDP)
+      new_hints.ai_protocol = IPPROTO_UDP;
+
+    if (hints->ai_family == MUSL_AF_INET)
+      new_hints.ai_family = AF_INET;
+    else if (hints->ai_family == MUSL_AF_INET6)
+      new_hints.ai_family = AF_INET6;
+  }
+
+  addrinfo* native_res = nullptr;
+  int result = getaddrinfo(node, service,
+                           reinterpret_cast<const struct addrinfo*>(&new_hints),
+                           &native_res);
+
+  struct addrinfo* ai = native_res;
+  struct musl_addrinfo* last_ai = nullptr;
+  *res = nullptr;
   while (ai != nullptr) {
     if (ai->ai_addr != nullptr) {
-      musl_sockaddr* musl_addr_ptr =
-          reinterpret_cast<musl_sockaddr*>(ai->ai_addr);
-      struct sockaddr* addr_ptr =
-          reinterpret_cast<struct sockaddr*>(ai->ai_addr);
-      uint8_t sa_family = addr_ptr->sa_family;
-      musl_addr_ptr->sa_family = sa_family;
+      addrinfo ai_copy = *ai;
+      struct musl_addrinfo* musl_ai =
+          (struct musl_addrinfo*)calloc(1, sizeof(struct musl_addrinfo));
+      musl_ai->ai_flags = ai_copy.ai_flags;
+      musl_ai->ai_family = ai_copy.ai_family;
+      musl_ai->ai_socktype = ai_copy.ai_socktype;
+      musl_ai->ai_protocol = ai_copy.ai_protocol;
+      musl_ai->ai_addrlen = ai_copy.ai_addrlen;
+      musl_ai->ai_addr =
+          (struct musl_sockaddr*)calloc(1, sizeof(struct musl_sockaddr));
+      musl_ai->ai_addr->sa_family = ai_copy.ai_addr->sa_family;
+      memcpy(musl_ai->ai_addr->sa_data, ai_copy.ai_addr->sa_data,
```

Comment:
In the Posix definition, I thought `char sa_data[]` means that it's a static array, "variable-length" means the size could differ on different platform(but it's still a static array). I wonder if there are platforms where the static size is shorter than 26 which could lead to reading out-of-bound. But let's follow your suggestion to avoid missing data, which is harder to catch.